-1

* **Build**:

I found compilation error when building: 

<pre>creating the shared library /build/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/BUILD/slc6_amd64_gcc481/external/evtgen/1.3.0-cms13/evtgen/1.3.0/lib/libEvtGenExternal.so
g++ -O2 -Wall -pthread -fPIC  ../tmp_ext/EvtGenExternal/EvtExternalGenFactory.o ../tmp_ext/EvtGenExternal/EvtPhotosEngine.o ../tmp_ext/EvtGenExternal/EvtExternalGenList.o ../tmp_ext/EvtGenExternal/EvtPHOTOS.o ../tmp_ext/EvtGenExternal/EvtPythia.o ../tmp_ext/EvtGenExternal/EvtPythiaEngine.o ../tmp_ext/EvtGenExternal/EvtPythia6CommandConverter.o ../tmp_ext/EvtGenExternal/EvtTauolaEngine.o ../tmp_ext/EvtGenExternal/EvtTauola.o -o /build/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/BUILD/slc6_amd64_gcc481/external/evtgen/1.3.0-cms13/evtgen/1.3.0/lib/libEvtGenExternal.so -shared -Wl,-soname,libEvtGenExternal.so -lgfortran -lgfortranbegin \
	-L/build/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/slc6_amd64_gcc481/external/hepmc/2.06.07-ddibom/lib -lHepMC -L/build/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/slc6_amd64_gcc481/external/photospp/3.61/lib -lPhotosCxxInterface -lPhotosFortran -L/build/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/slc6_amd64_gcc481/external/tauolapp/1.1.4-ddibom13/lib -lTauolaCxxInterface -lTauolaFortran -L/build/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/slc6_amd64_gcc481/external/pythia8/226-ddibom6/lib -lpythia8 -L/build/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/BUILD/slc6_amd64_gcc481/external/evtgen/1.3.0-cms13/evtgen/1.3.0/lib -lEvtGen
/build/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/slc6_amd64_gcc481/external/gcc/4.8.1/bin/../lib/gcc/x86_64-redhat-linux-gnu/4.8.1/../../../../x86_64-redhat-linux-gnu/bin/ld: cannot find -lPhotosCxxInterface
/build/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/slc6_amd64_gcc481/external/gcc/4.8.1/bin/../lib/gcc/x86_64-redhat-linux-gnu/4.8.1/../../../../x86_64-redhat-linux-gnu/bin/ld: cannot find -lPhotosFortran
collect2: error: ld returned 1 exit status
make[1]: *** [/build/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/BUILD/slc6_amd64_gcc481/external/evtgen/1.3.0-cms13/evtgen/1.3.0/lib/libEvtGenExternal.so] Error 1
rm ../tmp_ext/EvtGenExternal/EvtTauolaEngine.mk ../tmp_ext/EvtGenExternal/EvtPhotosEngine.mk ../tmp_ext/EvtGenExternal/EvtTauola.mk ../tmp_ext/EvtGenExternal/EvtPythiaEngine.mk ../tmp_ext/EvtGenExternal/EvtPHOTOS.mk ../tmp_ext/EvtGenExternal/EvtPythia.mk ../tmp_ext/EvtGenExternal/EvtPythia6CommandConverter.mk ../tmp_ext/EvtGenExternal/EvtExternalGenFactory.mk ../tmp_ext/EvtGenExternal/EvtExternalGenList.mk
make[1]: Leaving directory `/build/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/BUILD/slc6_amd64_gcc481/external/evtgen/1.3.0-cms13/evtgen/1.3.0/src'
make: *** [libext_shared] Error 2
error: Bad exit status from /build/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/tmp/rpm-tmp.B3tGg2 (%build)
</pre>

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a4055d/437/summary.html

--- a/GeneratorInterface/EvtGenInterface/plugins/EvtGen/EvtGenInterface.cc
+++ b/GeneratorInterface/EvtGenInterface/plugins/EvtGen/EvtGenInterface.cc
@@ -45,6 +45,10 @@
 #include "TString.h"
 #include <string>
 #include <stdlib.h> 
+#include <cstdio>
+
+#include "boost/filesystem.hpp"
+#include "boost/filesystem/path.hpp"
 
 using namespace gen;
 using namespace edm;
@@ -358,7 +362,10 @@ void EvtGenInterface::init(){
 
   if (fPSet->exists("user_decay_embedded")){
     std::vector<std::string> user_decay_lines = fPSet->getParameter<std::vector<std::string> >("user_decay_embedded");
-    std::string user_decay_tmp = std::tmpnam(nullptr);
+    auto tmp_dir = boost::filesystem::temp_directory_path();
+    tmp_dir += "%%%%-%%%%-%%%%-%%%%";
+    auto tmp_path = boost::filesystem::unique_path(tmp_dir);
+    std::string user_decay_tmp = std::string(tmp_path.c_str());
     FILE* tmpf = std::fopen(user_decay_tmp.c_str(), "w");
     if (!tmpf) {
         edm::LogError("EvtGenInterface::~EvtGenInterface") << "EvtGenInterface::init() fails when trying to open a temporary file for embedded user.dec. Terminating program ";

--- a/GeneratorInterface/EvtGenInterface/plugins/EvtGenUserModels/EvtLb2plnuLCSR.cpp
+++ b/GeneratorInterface/EvtGenInterface/plugins/EvtGenUserModels/EvtLb2plnuLCSR.cpp
@@ -107,7 +107,7 @@ void EvtLb2plnuLCSR::initProbMax() {
   }
   else
   {
-    report(ERROR,"EvtGen") << "Decay does not have Lb->p setting ProbMax = 0 "
+    EvtGenReport(EVTGEN_ERROR,"EvtGen") << "Decay does not have Lb->p setting ProbMax = 0 "
 			   <<endl;
     setProbMax(0.0);
   }
@@ -118,18 +118,18 @@ void EvtLb2plnuLCSR::init(){
   
   if (getNArg()!=4) {
 
-    report(ERROR,"EvtGen") << "EvtLb2plnuLCSR generator expected "
+    EvtGenReport(EVTGEN_ERROR,"EvtGen") << "EvtLb2plnuLCSR generator expected "
 			   << " 4 arguments but found:"<<getNArg()<<endl;
-    report(ERROR,"EvtGen") << "Will terminate execution!"<<endl;
+    EvtGenReport(EVTGEN_ERROR,"EvtGen") << "Will terminate execution!"<<endl;
     ::abort();
 
   }
 
   if ( getNDaug()!=3 ) {
-     report(ERROR,"EvtGen") 
+     EvtGenReport(EVTGEN_ERROR,"EvtGen") 
        << "Wrong number of daughters in EvtLb2plnu.cc " 
        << " 3 daughters expected but found: "<<getNDaug()<<endl;
-     report(ERROR,"EvtGen") << "Will terminate execution!"<<endl;
+     EvtGenReport(EVTGEN_ERROR,"EvtGen") << "Will terminate execution!"<<endl;
      ::abort();
   }
 
@@ -143,24 +143,24 @@ void EvtLb2plnuLCSR::init(){
   EvtSpinType::spintype neutrinotype=EvtPDL::getSpinType(getDaug(2));
 
   if ( parenttype != EvtSpinType::DIRAC ) {
-    report(ERROR,"EvtGen") << "EvtLb2plnuLCSR generator expected "
+    EvtGenReport(EVTGEN_ERROR,"EvtGen") << "EvtLb2plnuLCSR generator expected "
                            << " a DIRAC parent, found:"<<
                            EvtPDL::name(getParentId())<<endl;
-    report(ERROR,"EvtGen") << "Will terminate execution!"<<endl;
+    EvtGenReport(EVTGEN_ERROR,"EvtGen") << "Will terminate execution!"<<endl;
     ::abort();
   }
   if ( leptontype != EvtSpinType::DIRAC ) {
-    report(ERROR,"EvtGen") << "EvtLb2plnuLCSR generator expected "
+    EvtGenReport(EVTGEN_ERROR,"EvtGen") << "EvtLb2plnuLCSR generator expected "
                            << " a DIRAC 2nd daughter, found:"<<
                            EvtPDL::name(getDaug(1))<<endl;
-    report(ERROR,"EvtGen") << "Will terminate execution!"<<endl;
+    EvtGenReport(EVTGEN_ERROR,"EvtGen") << "Will terminate execution!"<<endl;
     ::abort();
   }
   if ( neutrinotype != EvtSpinType::NEUTRINO ) {
-    report(ERROR,"EvtGen") << "EvtLb2plnuLCSR generator expected "
+    EvtGenReport(EVTGEN_ERROR,"EvtGen") << "EvtLb2plnuLCSR generator expected "
                            << " a NEUTRINO 3rd daughter, found:"<<
                            EvtPDL::name(getDaug(2))<<endl;
-    report(ERROR,"EvtGen") << "Will terminate execution!"<<endl;
+    EvtGenReport(EVTGEN_ERROR,"EvtGen") << "Will terminate execution!"<<endl;
     ::abort();
   }
 
@@ -172,11 +172,11 @@ void EvtLb2plnuLCSR::init(){
     calcamp = new EvtSLBaryonAmp; 
   }
   else {
-    report(ERROR,"EvtGen") 
+    EvtGenReport(EVTGEN_ERROR,"EvtGen") 
       << "Wrong baryon spin type in EvtLb2plnuLCSR.cc " 
       << "Expected spin type " << EvtSpinType::DIRAC 
       << ", found spin type " << baryontype <<endl;
-    report(ERROR,"EvtGen") << "Will terminate execution!" <<endl;
+    EvtGenReport(EVTGEN_ERROR,"EvtGen") << "Will terminate execution!" <<endl;
      ::abort();
   }
   

--- a/GeneratorInterface/EvtGenInterface/plugins/EvtGenUserModels/EvtLb2plnuLCSRFF.cpp
+++ b/GeneratorInterface/EvtGenInterface/plugins/EvtGenUserModels/EvtLb2plnuLCSRFF.cpp
@@ -84,7 +84,7 @@ void EvtLb2plnuLCSRFF::getdiracff(EvtId parent, EvtId daught,
   }
   else 
   {
-  report(ERROR,"EvtGen") << "Only Lb -> p transitions allowed in EvtLb2plnuLCSRFF.\n";  
+  EvtGenReport(EVTGEN_ERROR,"EvtGen") << "Only Lb -> p transitions allowed in EvtLb2plnuLCSRFF.\n";  
   ::abort();
   }
 
@@ -97,14 +97,14 @@ void EvtLb2plnuLCSRFF::getraritaff( EvtId , EvtId ,
 				  double* , double* , double* , double*, 
 				  double* , double* , double* , double*  ) {
 
-  report(ERROR,"EvtGen") << "Not implemented :getraritaff in EvtLb2plnuLCSRFF.\n";  
+  EvtGenReport(EVTGEN_ERROR,"EvtGen") << "Not implemented :getraritaff in EvtLb2plnuLCSRFF.\n";  
   ::abort();
 
 }
 
 void EvtLb2plnuLCSRFF::getscalarff(EvtId, EvtId, double, double, double*, double*) {
 
-  report(ERROR,"EvtGen") << "Not implemented :getscalarff in EvtLb2plnuLCSRFF.\n";  
+  EvtGenReport(EVTGEN_ERROR,"EvtGen") << "Not implemented :getscalarff in EvtLb2plnuLCSRFF.\n";  
   ::abort();
 
 }
@@ -112,7 +112,7 @@ void EvtLb2plnuLCSRFF::getscalarff(EvtId, EvtId, double, double, double*, double
 void EvtLb2plnuLCSRFF::getvectorff(EvtId, EvtId, double, double, double*, double*,
 				 double*, double*) {
 
-  report(ERROR,"EvtGen") << "Not implemented :getvectorff in EvtLb2plnuLCSRFF.\n";  
+  EvtGenReport(EVTGEN_ERROR,"EvtGen") << "Not implemented :getvectorff in EvtLb2plnuLCSRFF.\n";  
   ::abort();
 
 }
@@ -120,7 +120,7 @@ void EvtLb2plnuLCSRFF::getvectorff(EvtId, EvtId, double, double, double*, double
 void EvtLb2plnuLCSRFF::gettensorff(EvtId, EvtId, double, double, double*, double*,
 				 double*, double*) {
 
-  report(ERROR,"EvtGen") << "Not implemented :gettensorff in EvtLb2plnuLCSRFF.\n";  
+  EvtGenReport(EVTGEN_ERROR,"EvtGen") << "Not implemented :gettensorff in EvtLb2plnuLCSRFF.\n";  
   ::abort();
 
 }
@@ -128,7 +128,7 @@ void EvtLb2plnuLCSRFF::gettensorff(EvtId, EvtId, double, double, double*, double
 void EvtLb2plnuLCSRFF::getbaryonff(EvtId, EvtId, double, double, double*, 
 				 double*, double*, double*){
   
-  report(ERROR,"EvtGen") << "Not implemented :getbaryonff in EvtLb2plnuLCSRFF.\n";  
+  EvtGenReport(EVTGEN_ERROR,"EvtGen") << "Not implemented :getbaryonff in EvtLb2plnuLCSRFF.\n";  
   ::abort();
 
 }

--- a/GeneratorInterface/EvtGenInterface/plugins/EvtGenUserModels/EvtSLBaryonAmp.cpp
+++ b/GeneratorInterface/EvtGenInterface/plugins/EvtGenUserModels/EvtSLBaryonAmp.cpp
@@ -152,7 +152,7 @@ void EvtSLBaryonAmp::CalcAmp( EvtParticle *parent,
 			    parent->getDaug(1)->spParent(1));
     }
     else{
-      report(ERROR,"EvtGen") << "Wrong lepton number"<<endl;
+      EvtGenReport(EVTGEN_ERROR,"EvtGen") << "Wrong lepton number"<<endl;
     }
   }
 
@@ -477,7 +477,7 @@ void EvtSLBaryonAmp::CalcAmp(EvtParticle *parent,
 			    parent->getDaug(1)->spParent(1));
       
     } else {
-      report(ERROR,"EvtGen")<< "Wrong lepton number \n";
+      EvtGenReport(EVTGEN_ERROR,"EvtGen")<< "Wrong lepton number \n";
       ::abort();
     }
 
@@ -555,7 +555,7 @@ void EvtSLBaryonAmp::CalcAmp(EvtParticle *parent,
     }
 
     else {
-      report(ERROR,"EvtGen") << "Dirac semilep. baryon current " 
+      EvtGenReport(EVTGEN_ERROR,"EvtGen") << "Dirac semilep. baryon current " 
 			     << "not implemented for this decay sequence." 
 			     << std::endl;
       ::abort();
@@ -609,7 +609,7 @@ void EvtSLBaryonAmp::CalcAmp(EvtParticle *parent,
 			    parent->getDaug(1)->spParent(1));
       
     } else {
-      report(ERROR,"EvtGen")<< "Wrong lepton number \n";
+      EvtGenReport(EVTGEN_ERROR,"EvtGen")<< "Wrong lepton number \n";
     }
       
     //  Baryon Current
@@ -647,7 +647,7 @@ void EvtSLBaryonAmp::CalcAmp(EvtParticle *parent,
       pflag = 2;
     }
     else {
-      report(ERROR,"EvtGen") << "Rarita-Schwinger semilep. baryon current " 
+      EvtGenReport(EVTGEN_ERROR,"EvtGen") << "Rarita-Schwinger semilep. baryon current " 
 			     << "not implemented for this decay sequence." 
 			     << std::endl;
       ::abort();

--- a/GeneratorInterface/ExternalDecays/src/ExternalDecayDriver.cc
+++ b/GeneratorInterface/ExternalDecays/src/ExternalDecayDriver.cc
@@ -61,9 +61,9 @@ ExternalDecayDriver::ExternalDecayDriver( const ParameterSet& pset )
 	exSharedResources.emplace_back(edm::SharedResourceNames::kPhotos);
       }
     }
-    else if (curSet == "Photospp355" ){
+    else if (curSet == "Photospp" || curSet == "Photospp356" ){
       if ( !fPhotosInterface ){
-        fPhotosInterface = (PhotosInterfaceBase*)(PhotosFactory::get()->create("Photospp355", pset.getUntrackedParameter< ParameterSet>(curSet)));
+        fPhotosInterface = (PhotosInterfaceBase*)(PhotosFactory::get()->create("Photospp356", pset.getUntrackedParameter< ParameterSet>(curSet)));
         exSharedResources.emplace_back(edm::SharedResourceNames::kPhotos);
       }
     }

--- a/GeneratorInterface/PhotosInterface/plugins/PhotosppInterface.cc
+++ b/GeneratorInterface/PhotosInterface/plugins/PhotosppInterface.cc
@@ -25,8 +25,16 @@ PhotosppInterface::PhotosppInterface( const edm::ParameterSet& pset)
     fIsInitialized(false),
     fPSet(0)
 {
-   fSpecialSettings.push_back("QED-brem-off:all");
-   fPSet = new ParameterSet(pset);
+  // add ability to keep brem from hadronizer and only modify specific channels 10/27/2014
+  bool UseHadronizerQEDBrem=false;
+  fPSet = new ParameterSet(pset);
+  std::vector<std::string> par = fPSet->getParameter< std::vector<std::string> >("parameterSets");
+  for (unsigned int ip=0; ip<par.size(); ++ip ){
+    std::string curSet = par[ip];
+    // Physics settings
+    if(curSet=="UseHadronizerQEDBrem") UseHadronizerQEDBrem=true;
+  }
+  if(!UseHadronizerQEDBrem)fSpecialSettings.push_back("QED-brem-off:all");
 }
 
 void PhotosppInterface::setRandomEngine(CLHEP::HepRandomEngine* decayRandomEngine){fRandomEngine=decayRandomEngine;}
@@ -209,4 +217,4 @@ double PhotosppInterface::flat(){
 
 void PhotosppInterface::statistics(){Photospp::Photos::iniInfo();}
 
-DEFINE_EDM_PLUGIN(PhotosFactory, gen::PhotosppInterface, "Photospp355");
+DEFINE_EDM_PLUGIN(PhotosFactory, gen::PhotosppInterface, "Photospp356");

--- a/GeneratorInterface/PhotosInterface/test/DY_7TeV_pythia8.py
+++ b/GeneratorInterface/PhotosInterface/test/DY_7TeV_pythia8.py
@@ -146,7 +146,7 @@ process.genParticles = cms.EDProducer("GenParticleProducer",
 )
 process.printTree1 = cms.EDAnalyzer("ParticleListDrawer",
     src = cms.InputTag("genParticles"),
-    maxEventsToPrint  = cms.untracked.int32(10)
+    maxEventsToPrint  = cms.untracked.int32(0)
 )
 
 process.path = cms.Path(process.generator*process.genParticles*process.printTree1)

